codalab/apps/newsletter/tests/test_newsletter_opt_in.py

```diff
@@ -0,0 +1,133 @@
+import hashlib
+import json
+
+import requests
+from django.test import TestCase
+from django.contrib.auth import get_user_model
+
+from apps.newsletter.models import NewsletterUser
+from codalab import settings
+
+User = get_user_model()
+
+
+class NewsletterOptIn(TestCase):
```

Comment:
We should definitely be `mock.patch`ing mailchimp here.

Reply:
Mock added to the tests

Reply:
should also add a few more tests:
```python 
def test_needs_retry_on_patch_fail(self):
    self.user.newsletter_opt_in = True
    # clear the detail so we don't send a post request
    mock_post, mock_patch = self.mock_save_user(patch_status_code=400, post_status_code=400, detail='')
    assert mock_patch.called
    assert not mock_post.called
    assert NewsletterSubscription.objects.filter(email=self.user.email).exists()
    newsletter_user = NewsletterSubscription.objects.get(email=self.user.email)
    assert newsletter_user.needs_retry
    assert newsletter_user.subscription_active

def test_needs_retry_on_patch_fail(self):
    self.user.newsletter_opt_in = True
    mock_post, mock_patch = self.mock_save_user(patch_status_code=400, post_status_code=400)
    assert mock_patch.called
    assert mock_post.called
    assert NewsletterSubscription.objects.filter(email=self.user.email).exists()
    newsletter_user = NewsletterSubscription.objects.get(email=self.user.email)
    assert newsletter_user.needs_retry
    assert newsletter_user.subscription_active

def test_needs_retry_on_unsub_fail(self):
    self.user.newsletter_opt_in = False
    mock_post, mock_patch = self.mock_save_user(patch_status_code=400, post_status_code=400)
    assert mock_patch.called
    assert mock_post.called
    assert NewsletterSubscription.objects.filter(email=self.user.email).exists()
    newsletter_user = NewsletterSubscription.objects.get(email=self.user.email)
    assert newsletter_user.needs_retry
    assert not newsletter_user.subscription_active
```

I think these all should work. but I didn't wanna take the time to set shit up on my end and actually run the tests, but I think the logic is solid.

Reply:
Also, looks like there are a couple tests failing on circleCI